source/extensions/filters/http/jwt_authn/filter.cc

```diff
@@ -0,0 +1,97 @@
+#include "extensions/filters/http/jwt_authn/filter.h"
+
+#include "common/http/utility.h"
+
+using ::google::jwt_verify::Status;
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace JwtAuthn {
+
+Filter::Filter(FilterConfigSharedPtr config) : config_(config) {}
+
+void Filter::onDestroy() {
+  ENVOY_LOG(debug, "Called Filter : {}", __func__);
+  if (auth_) {
+    auth_->onDestroy();
+  }
+}
+
+Http::FilterHeadersStatus Filter::decodeHeaders(Http::HeaderMap& headers, bool) {
+  ENVOY_LOG(debug, "Called Filter : {}", __func__);
+
+  // TODO: support per-route config.
+  // store_factory can be created in route::per_filter_config
+
+  // For now, only use config from filter.
+  if (!config_) {
+    return Http::FilterHeadersStatus::Continue;
+  }
+  auth_ = Authenticator::create(config_);
+
+  // Remove headers configured to pass payload
+  auth_->sanitizePayloadHeaders(headers);
+
+  state_ = Calling;
+  stopped_ = false;
+
+  // Verify the JWT token, onComplete() will be called when completed.
+  auth_->verify(headers, this);
+  if (state_ == Complete) {
+    return Http::FilterHeadersStatus::Continue;
+  }
+  ENVOY_LOG(debug, "Called Filter : {} Stop", __func__);
+  stopped_ = true;
+  return Http::FilterHeadersStatus::StopIteration;
+}
+
+void Filter::onComplete(const Status& status) {
+  ENVOY_LOG(debug, "Called Filter : check complete {}", int(status));
+  // This stream has been reset, abort the callback.
+  if (state_ == Responded) {
+    return;
+  }
+  if (status != Status::Ok) {
+    config_->stats().denied_.inc();
+    state_ = Responded;
+    // verification failed
+    Http::Code code = Http::Code::Unauthorized;
+    // return failure reason as message body
+    Http::Utility::sendLocalReply(false, *decoder_callbacks_, false, code,
+                                  ::google::jwt_verify::getStatusString(status));
```

Comment:
@saumoh can you comment on how this relates to the state machine and handling done for ext authz?